### ISSUE TITLE
codes problem

### ISSUE THREAD

**Issue by klyao** [williamleif/GraphSAGE#39]
hi, I don't understand the meaning of the second 'for' loop in

models.py
--class SampleAndAggregate(GeneralizedModel):
----def aggregate(...)

It seems that it is not consistent with pseudo-code in your paper.

(By the way, your code is a little hard to follow T_T)

**Comment by williamleif** [comment]
Hi yes, this part of the code is extrement hard to parse, though it is doing what it needs to.. The main complicating thing (which made directly implementing the psuedo-code exactly as written impossible in Tensorflow) was TensorFlow's (at the time) lack of support for dynamic compute graphs. The concatenation in line 5 of algorithm 1 also really complicates implementation in TensorFlow. 

For the sake of my sanity and others' understanding, I made an easier to understand reference implementation in PyTorch if you're interested: https://github.com/williamleif/graphsage-simple 

I did not actually write those lines of code, but I'll try to explain a bit, though it is *really* confusing (thus why I made the easier-to-read pytorch version): 

The reason for the two loops is that to get the final "output" representation of a node, we need representations of that node at each of the preceding layers (in the notation of Algorithm 1, to get the iteration k=K representation of a node, we need its representations for iterations 1...K_1). So basically the first iteration of the inner-loop (when layer=0 and hop=0) corresponds to getting the iteration k=1 representations for the set of nodes that we want to have embeddings for at the final "layer"/iteration k=K. And the inner-loop when layer=0, hop=1 computers the iteration k=1 embeddings for nodes that are needed at iteration k=2, and the layer=0, hop=2 inner-loop computes the embeddings for the nodes that are only used at iteration k=1. This is horrendously confusing imo haha T_T, but it works and avoids unnecessary computations compared to more naive TensorFlow implementations.
